tests/realtikvtest/txntest/pipelineddml_test.go

```diff
@@ -0,0 +1,440 @@
+	tk.MustExec("insert ignore into _t select * from t")
+	require.Equal(t, tk.Session().AffectedRows(), uint64(9989))
+	tk.MustQuery("select count(1) from _t").Check(testkit.Rows("10000"))
+	tk.MustQuery("select * from _t where a in (0, 999, 1999, 2999, 3999, 4999, 5999, 6999, 7999, 8999, 9999)").Sort().
+		Check(testkit.Rows("0 -1", "1999 -1", "2999 -1", "3999 -1", "4999 -1", "5999 -1", "6999 -1", "7999 -1", "8999 -1", "999 -1", "9999 -1"))
+}
+
+func TestPipelinedDMLInsertOnDuplicateKeyUpdate(t *testing.T) {
+	store := realtikvtest.CreateMockStoreAndSetup(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	prepareData(tk)
+	tk.MustExec("insert into _t values(0, -1), (999, -1), (1999, -1), (2999, -1), (3999, -1), (4999, -1), (5999, -1), (6999, -1), (7999, -1), (8999, -1), (9999, -1)")
+	tk.MustExec("insert into _t select * from t on duplicate key update b = values(b)")
+	require.Equal(t, tk.Session().AffectedRows(), uint64(10011))
+	compareTables(t, tk, "t", "_t")
+}
+
+func TestPipelinedDMLInsertOnDuplicateKeyUpdateInTxn(t *testing.T) {
+	require.Nil(t, failpoint.Enable("tikvclient/pipelinedMemDBMinFlushKeys", `return(10)`))
+	require.Nil(t, failpoint.Enable("tikvclient/pipelinedMemDBMinFlushSize", `return(100)`))
+	require.Nil(t, failpoint.Enable("tikvclient/pipelinedMemDBForceFlushSizeThreshold", `return(10240)`))
+	defer func() {
+		require.Nil(t, failpoint.Disable("tikvclient/pipelinedMemDBMinFlushKeys"))
+		require.Nil(t, failpoint.Disable("tikvclient/pipelinedMemDBMinFlushSize"))
+		require.Nil(t, failpoint.Disable("tikvclient/pipelinedMemDBForceFlushSizeThreshold"))
+	}()
+	store := realtikvtest.CreateMockStoreAndSetup(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t1")
+	tk.MustExec("create table t1 (a int, b int, c varchar(128), unique index idx(b))")
+	cnt := 2000
+	values := bytes.NewBuffer(make([]byte, 0, 10240))
+	for i := 0; i < cnt; i++ {
+		if i > 0 {
+			values.WriteString(", ")
+		}
+		if i == 1500 {
+			values.WriteString(fmt.Sprintf("(%d, %d, 'abcdefghijklmnopqrstuvwxyz1234567890,.?+-=_!@#$&*()_+')", i, 250))
+		} else {
+			values.WriteString(fmt.Sprintf("(%d, %d, 'abcdefghijklmnopqrstuvwxyz1234567890,.?+-=_!@#$&*()_+')", i, i))
+		}
+	}
+	tk.MustExec("set session tidb_dml_type = bulk")
+	// Test insert meet duplicate key error.
+	tk.MustGetErrMsg("insert into t1 values "+values.String(), "[kv:1062]Duplicate entry '250' for key 't1.idx'")
+	tk.MustQuery("select count(*) from t1").Check(testkit.Rows("0"))
+	// Test insert ignore
+	tk.MustExec("insert ignore into t1 values " + values.String())
+	tk.MustQuery("show warnings").Check(testkit.Rows("Warning 1062 Duplicate entry '250' for key 't1.idx'"))
+	tk.MustQuery("select count(*) from t1").Check(testkit.Rows("1999"))
+	tk.MustQuery("select a, b from t1 where a in (250, 1500)").Check(testkit.Rows("250 250"))
+	// Test insert on duplicate key update.
+	if !*realtikvtest.WithRealTiKV {
```

Comment:
```suggestion
	if !*realtikvtest.WithRealTiKV {
		// TODO: fix me. skip for real TiKV because now we have assertion issue.
		// Test replace into.
		tk.MustExec("delete from t1")
		tk.MustExec("replace into t1 values " + values.String())
		tk.MustQuery("select count(*) from t1").Check(testkit.Rows("1999"))
		tk.MustQuery("select a, b from t1 where a in (250, 1500)").Check(testkit.Rows("1500 250"))
		// Test insert on duplicate key update.
```

Reply:
add test for replace into